enhanced tabs

exeea opened the pull request:
Reimplementation of the tab system to work with multiple units simultaneously. 
The main enhancements include:

- Persistent Action Bar: The "new unit" and "load unit" buttons remain visible and accessible regardless of how many tabs are open. Two positioning modes are supported: trailing mode (default, after the last tab) or fixed mode (right edge of the tab pane)
- Detachable Tabs: Units can be popped out into separate windows
- Tab Sorting: Tabs can be reordered via drag and drop
- Visual Feedback: Ghost tab visualization during drag operations provides clear feedback

These features have been implemented in two places:
- The main unit tabs in MegaMekLabTabbedUI
- The detail tabs within each unit editor (Structure, Equipment, Quirks, Preview, etc.)

Sleet01 commented:
I'll review this in depth when I'm back at the computer.
Can you address the CodeQL alerts, either with fixes or with comments regarding why they can be ignored?

exeea commented on megameklab/src/megameklab/ui/util/EnhancedTabbedPane.java:
```diff
@@ -0,0 +1,1299 @@
+
+        // Notify listeners before detaching
+        fireTabDetaching(tabIndex, title, component);
+
```
title is already removed

exeea commented on megameklab/src/megameklab/ui/util/EnhancedTabbedPane.java:
```diff
@@ -0,0 +1,1299 @@
+         * @param tabIndex  The index of the tab being detached
+         * @param component The component in the tab
+         */
+        default void onTabDetaching(int tabIndex, Component component) {
```
is a core argument of the interface for listening tab state changes

exeea commented on megameklab/src/megameklab/ui/util/EnhancedTabbedPane.java:
```diff
@@ -0,0 +1,1299 @@
+         * @param tabIndex  The index where the tab was inserted
+         * @param component The component that was reattached
+         */
+        default void onTabReattached(int tabIndex, Component component) {
```
is a core argument of the interface for listening tab state changes

exeea commented on megameklab/src/megameklab/ui/util/EnhancedTabbedPane.java:
```diff
@@ -0,0 +1,1299 @@
+         * @param tabIndex  The index of the tab being detached
+         * @param component The component in the tab
+         */
+        default void onTabDetaching(int tabIndex, Component component) {
```
is a core argument of the interface for listening tab state changes

exeea commented on megameklab/src/megameklab/ui/util/EnhancedTabbedPane.java:
```diff
@@ -0,0 +1,1299 @@
+         * @param tabIndex  The index where the tab was inserted
+         * @param component The component that was reattached
+         */
+        default void onTabReattached(int tabIndex, Component component) {
```
is a core argument of the interface for listening tab state changes

exeea commented:
I fixed the open codeql issues. Some of them I marked as resolved because they are false positives.

Sleet01 commented on megameklab/src/megameklab/ui/MegaMekLabTabbedUI.java:
```diff
@@ -180,7 +342,8 @@ private void newUnit(long type, boolean primitive, boolean industrial) {
         editors.set(tabs.getSelectedIndex(), newUi);
         tabs.setComponentAt(tabs.getSelectedIndex(), newUi.getContentPane());
-        tabs.setTabComponentAt(tabs.getSelectedIndex(), new EditorTab(newUi.getEntity().getDisplayName(), newUi));
+        // tabs.setTabComponentAt(tabs.getSelectedIndex(), new
```
Remove.

Sleet01 commented:
This is very impressive!  Overall the code looks good, but there are a few rough edges with the UX:

1. Tearing away the only tab leaves a new window with no tab controls, and the old window with no contents except for the two tiny tab buttons.  Additionally, closing the original (now blank) window will close the torn-off window as well.
2. It's not possible to drag a tab to the right of the two tab control buttons ("new" and "open") as this causes the tab to be torn off as a new window.
3. Closing a window to return the tab to the tabbed view is not very intuitive.

At the very least I'd like to see #1 addressed, probably by just disabling the tab drag functionality if there is only one open tab.

Finally, we are very close to the 0.50.04 release and should not merge any non-bug-fix PRs at this point.

exeea commented:
1) ok.

2) I can put a constrain that doesn't let undock for the whole length of the tab pane/window.

3) I can try to put a "dock" icon in the caption. Need to figure out if creates OS compatibility issues.


I can also offer the option that, when a window is floated, it also creates a full tabbed container with his own new/load buttons. Basically like Chrome&friends works, so any unit's tabbed window can act as a dock and I can enable dock/undock between them.

Sleet01 commented:
> 1. ok.
> 2. I can put a constrain that doesn't let undock for the whole length of the tab pane/window.
> 3. I can try to put a "dock" icon in the caption. Need to figure out if creates OS compatibility issues.
> 
> I can also offer the option that, when a window is floated, it also creates a full tabbed container with his own new/load buttons. Basically like Chrome&friends works, so any unit's tabbed window can act as a dock and I can enable dock/undock between them.

I was thinking of Chrome, but I was hesitant to ask - that seems like a ton of extra tracking work.
TBH, 1. and 2. were the things that got me; I figured the close -> dock would work but it wasn't the first thing that sprang to mind.
Again, thanks for all the hard work!

pavelbraginskiy commented:
Working like how browsers work is definitely the best possible scenario, if it's doable.
"Bonus points" would be if you could open two instances of MML and pass tabs between them.

exeea commented:
for "3" I implemented a drag-to-dock functionality.
For the units' tabs (structure/equipment/etc...) the docking area is the whole tabs area (full width).
For the main unit tab the docking area is the visible tabs area. (this is required for the future implementation of multi-windows with shared docking)

Additionally I added that the undock happens only on mouse release, this is useful because:
-  makes easier to reorder the tabs without unwanted undocking
-  in later developments will allow to move tabs between megameklab tabbed windows without jarring undock/dock animations in the middle 

(as always, remains available the option to quickly undock with right-click on a tab)

exeea commented:
> "Bonus points" would be if you could open two instances of MML and pass tabs between them - could be done by serializing entities between processes, or by the second MML process just requesting the first one to open a new window and terminating itself.

But is there an user case for that scenario? 
Considering that we might implement that:
- by undocking an unit's tab, you get a full MegaMekLabTabbedUI instance with it
- tabs can be moved between multiple instances of MegaMekLabTabbedUI

Sleet01 reviewed (approved):
Looks great!
